treesit-auto.el

```diff
@@ -387,39 +364,51 @@ how to modify the behavior of this function."
               (y-or-n-p "Install missing grammars? "))
       (mapcar 'treesit-install-language-grammar to-install))))
 
-(defun treesit-auto--install-language-grammar-wrapper (&rest _r)
-  "Run `treesit-auto-apply-remap' after `treesit-install-language-grammar'."
-  (treesit-auto--build-alists))
-
-(defun treesit-auto--setup ()
-  "Set up `global-treesit-auto-mode'."
-  (setq treesit-auto--original-major-mode-remap-alist (purecopy major-mode-remap-alist))
-  (setq treesit-auto--original-language-source-alist (purecopy treesit-language-source-alist))
-  (treesit-auto--build-alists)
-  (dolist (elt (mapcar 'car treesit-auto-mode-lang-alist))
-    (add-hook (intern (concat (symbol-name elt) "-hook"))
-              #'treesit-auto--maybe-install-grammar))
-  (advice-add 'treesit-install-language-grammar
-	      :after #'treesit-auto--install-language-grammar-wrapper))
-
-(defun treesit-auto--teardown ()
-  "Undo any change made by `global-treesit-auto-mode'."
-  (advice-remove 'treesit-install-language-grammar
-                 #'treesit-auto--install-language-grammar-wrapper)
-  (dolist (elt (mapcar 'car treesit-auto-mode-lang-alist))
-    (remove-hook (intern (concat (symbol-name elt) "-hook"))
-                 #'treesit-auto--maybe-install-grammar))
-  (setq major-mode-remap-alist (purecopy treesit-auto--original-major-mode-remap-alist))
-  (setq treesit-language-source-alist (purecopy treesit-auto--original-language-source-alist)))
-
 ;;;###autoload
-(define-minor-mode global-treesit-auto-mode
+(define-minor-mode treesit-auto-mode
   "Toggle `global-treesit-auto-mode'."
+  :group 'treesit)
+
+;;;###autoload
+(define-globalized-minor-mode global-treesit-auto-mode treesit-auto-mode
+  treesit-auto--on
   :group 'treesit
-  :global 't
+  :predicate
+  ;; allow global mode to activate only on recipe modes,
+  ;; but also allow to activate on remap and ts-modes
+  ;; in case only the ts-mode is available.
+  ;; non emacs core ts-modes might autoload and would be
+  ;; nice to also prompt for grammar installation
+  (let ((modes '()))
+    (cl-loop for recipe in treesit-auto-recipe-list
+             do (add-to-list 'modes (treesit-auto-recipe-ts-mode recipe))
+             do (dolist (mode (ensure-list (treesit-auto-recipe-remap recipe)))
+                  (add-to-list 'modes mode))
```

Comment:
Flymake error: `'add-to-list' can't use lexical var 'modes'; use 'push' or 'cl-pushnew'`

```suggestion
             do (dolist (mode (ensure-list (treesit-auto-recipe-remap recipe)))
                  (push mode 'modes))
```
```